### ISSUE TITLE
Contributions are not appreciated

### ISSUE THREAD

**Issue by dskloet** [bitpay/bitcore#1252]
I think it's great that BitPay is offering this great library to the community. But if you would like the community to contribute back to you, I have some advice: Treat your contributors well. Be polite, thank them for their contributions, answers their question and don't make them feel like a useless piece of shit.

In #1226 I simplified some code. The pull request was merged without any comment about the code being too much simplified. Why would it? Simple code is best code, right?

But in #1241 I fixed a bug (#1236), touching the same code as #1226 but otherwise unrelated to it. But braydonf held that pull request hostage by demanding that I undo the simplification made in #1226, making the code worse. My bug fix was simple and there was no reason to make the code more complicated. His request was unrelated to the bug fix and would make the code worse so I asked why he was asking me to undo my own improvements. But instead of answering my questions, Braydon took my code, made it worse and made his own pull request in #1248, without even informing me.

If you don't want people to contribute, I suggest you update your documentation to say so. But please don't insult people who are trying to help, by ignoring them and undoing their improvements "behind their backs" (I know it's all open, but still I had to figure it out myself instead of being informed directly).

**Comment by gordonwritescode** [comment]
Hi @dskloet,

Let me start by saying that your contributions to the project are **most** welcome and *very appreciated*. I am sorry that you are frustrated with the issues you outlined above. Let me try and answer some of your questions and provide a little bit of context.

I think that the source of some of this disdain (like most disputes), can be chalked up to a simple lack of communication. On #1226 it was mentioned that the order of some checks needed to be reviewed due to another issue #1236. I think if we had better communicated that the importance of resolving bugs outweighs the necessity to re-factor (especially when dealing with code that is being analysed for both purposes simultaneously), then perhaps this could have been resolved more amicably. On a more introspective note, I understand that it can be frustrating to be have your work undone (especially without being notified or told why), but it's important to accept that code will come and code will go and, in doing so, is not a reflection of the author(s), but a reflection of the needs of the project (almost as an autonomous entity of it's own).

Regarding a question you posted on #1241:

> It would also help if I had a better idea about how all of this works. Are you guys BitPay employees? Is Bitcore your full time job or a side project? Do you require extra time to discuss some pull requests on another channel that I can't see to make sure you agree about it? The whole process seems pretty opaque and contributing would be more fun if I could understand it better. Or maybe there isn't more too it than I can see but that would be good to know as well.

Most of the "top" contributors (per GitHub's charts) are currently, or have at some point, been employed by BitPay. However, @ryanxcharles, @maraoz, @eordano, @yemel, @chjj unfortunately are **not** employed by BitPay anymore. While this most certainly does not bar them from actively contributing, before they had (in varying degrees) a dedicated amount of time every day to this project.

As it currently stands, the only listed contributor who is employed by BitPay and allocates a given amount of time every day to this project is @braydonf. This means that your contributions are more important than ever right now, while we are trying to fill the gaps left by the aforementioned, dearly missed, comrades.

I think that what we need now is better transparency regarding decisions, priorities, and the future milestones for Bitcore. It is certainly no secret that Bitcore supports many of BitPay's efforts, both other free projects and tooling we use internally. Providing better transparency and context with these other efforts may serve well to improve collaboration between our engineers and the community. Hopefully that means releasing more of our work as free software.

Here are a few things I propose that may help:

* Remove our internal private #bitcore IRC channel and use a public one **exclusively**
* Make the source for other projects that heavily rely on Bitcore, like ChainDB, public so that issues can be cross-referenced and provide better context to the community
* Begin fostering better relationships with the community by giving frequent contributors write access to the repository

All in all, the last feeling we want from the community is that we are indifferent to your hard work. We recognize there are areas where we can improve and are making an active effort to do so.

**Comment by gordonwritescode** [comment]
@dskloet, in an effort to convert this thread into an opportunity for constructive feedback, I'd like to suggest changing the title to something like:

"Improve Project Transparency" or "Better Feedback Loop for Contributors"

**Comment by dskloet** [comment]
Hi @gordonwritescode,

Thanks for your response.
Yes, communication is what was sorely lacking.

You mention that fixing bugs is the highest priority. Why then, I wonder, did @braydonf demand that I add unrelated refactorings to my pull request that fixed a bug with minimal changes and increased test coverage? It would make more sense to me to merge the fix first and discuss unrelated "improvements" separately.

I completely accept that code comes and goes. Honestly, I think changes that delete code are the best. Even if that code is my own. But in this case, the work that was undone was a change to simplify the code so undoing it made the code worse. That's what is so frustrating about it.

Thanks for listening,

**Comment by maraoz** [comment]
@dskloet Your contributions are very appreciated, I'm sorry if you felt left out of the decision process. Truth is, bitcore is a community project, but financially supported by (owned by) BitPay. In most technical discussions, they'll have the last word, and that's fair, given they use bitcore for many of their core components/projects. 

Regarding the issue with https://github.com/bitpay/bitcore/pull/1241, it was an unfortunate confusion. I do agree with @dskloet in that PR reviews shouldn't ask for fixes not related to the work done in that PR. It's frustrating to have one's contribution blocked by something unrelated to it. So, I suggest that in the future, we all try to adhere to the rule of "only review the changes proposed in that PR".

I think giving write access to frequent external contributors is a great idea. I'd be happy to help with that :)

**Comment by dskloet** [comment]
Of course BitPay has the last word. They can ruin the code all they want ;-).
I'm just saying that if you value contributors, there is more to consider than just who is the boss.

And even if it was so important to undo my simplification (which I still don't understand), it could have been done after merging my PR which fixed the bug, instead of copying my PR into another PR to merge it together with the other "improvements". That feels like stealing my credit.

**Comment by dskloet** [comment]
There were some good suggestions mentioned here.
Did any of them get implemented in the meantime?

**Comment by Levino** [comment]
@maraoz Has Bitpay an actual interest in including the community? Currently there are 84 open issues, dating back to February 2014. Do you plan to clean this up any time soon?

**Comment by maraoz** [comment]
@Levino I'm no longer a BitPay employee, but continue to use and collaborate on bitcore. I know they're using it too, so... yes. You're welcome to collaborate too!

**Comment by braydonf** [comment]
Many bugs, features and improvements would not have happened without community involvement. Contributions from the community are appreciated.

**Comment by bookchin** [comment]
@braydonf well it's about time. :)
